A faster prettier voxblox

ZacharyTaylor opened the pull request:
Almost certainly not ready for merging, but I will probably refer to the pr when talking about some of the more questionable hacks in it.

![peek 2017-08-07 23-05](https://user-images.githubusercontent.com/730680/29046058-212d252a-7bc6-11e7-9b21-6e5f497b21a8.gif)

Long story short this pull request makes the fast integrator 5x slower, but a hell of a lot prettier. It also threads it and the concurrency is good enough that we are only a bit slower then before.

Meshing is also over 10x faster as we were previously computing unneeded normals.
Most publishing can now be disabled as this bottlenecks high res maps.

Oh my tests we now get a reliable 5Hz on 2cm voxels with voxel carving on.

ZacharyTaylor commented:
The branch is now in a state where it is ready for review.

New in this pull request:
- Ray traces blocks to allow preallocation of all memory needed, allowing concurrent access to the block hash map.
- Uses an array of 4096 mutexes to allow finer grained locking of voxels. This locking happens automatically when updateVoxel is called.
- Bringing together some of the common functions from the merged, fast and simple integrators to reduce code double up.
- Threading of the simple and fast integrator and changing the merging integrator to use the same threading strategy as the others.
- Removed second computation of normals in meshing, as this is done for free in the marching cubes part.
- Added flags for disabling some published information.

helenol commented:
@ZacharyTaylor when you get the chance, can you merge master into this branch? (Has conflicts)
Will do the review later today.

mfehr commented on voxblox/include/voxblox/utils/approx_hash_array.h:
```diff
@@ -0,0 +1,153 @@
+
+namespace voxblox {
+template <size_t unmasked_bits_, typename StoredElement>
+class ApproxHashArray {
```
How confident are you that this is all correct? Because for such fundamental containers it would be nice to have at least some basic unit testing :)

mfehr commented on voxblox_ros/src/voxblox_node.cc:
```diff
@@ -191,15 +197,25 @@ VoxbloxNode::VoxbloxNode(const ros::NodeHandle& nh,
-  tsdf_pointcloud_pub_ = nh_private_.advertise<pcl::PointCloud<pcl::PointXYZI>>(
-      "tsdf_pointcloud", 1, true);
+
+  if (publish_tsdf_info_) {
```
Very nice, when I recently used voxblox I was amazed by how much stuff it publishes :)

helenol replied:
😉

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+
+  // Can be used to pre-allocate all the blocks needed by the integrator. This
+  // allows thread-safe lockless block map access
+  void allocateBlocks(const Transformation& T_G_C, const Pointcloud& points_C,
```
As a small sidenote re: transforming, you do need points_C for the integration as well; not just points_G. While I in general agree with trying not to duplicate work/etc., I'm not 100% sure how huge of a difference at least the pre-transform will make?

ZacharyTaylor replied:
I agree the point getter is a bit ugly here. The idea was to place the clearing ray logic inside it so it is all in one place, but I think in practice it made things uglier.

mfehr replied:
@helenol yes we need both points_C and _G, but I think we should not transform them twice if it can be avoided and regarding the PointGetter, I do see how it keeps the integrator functions very short and simple, but IMO they became harder to understand, especially the data flow and it introduced this redundancy regarding the transformation. But let's discuss it when @ZacharyTaylor  is back, unless someone is in a hurry to merge this?

ZacharyTaylor commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -17,8 +18,11 @@
+
 namespace voxblox {
 
 class TsdfIntegratorBase {
```
Agreed will split it up

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+    }
+  }
+
+  inline TsdfVoxel* getVoxel(const VoxelIndex& global_voxel_idx,
```
:+1:
Maybe "getVoxelFromGlobalVoxelIndex" as well or whatever

ZacharyTaylor commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+    }
+
+    bool getNextPoint(size_t* point_idx, Point* point_G, bool* clearing) {
+      while (atomic_idx_.load() < points_C_.size()) {
```
If the point is too close or too far away it needs to go back and try get the next point, until it hits the end or finds a valid one

ZacharyTaylor commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+
+    bool getNextPoint(size_t* point_idx, Point* point_G, bool* clearing) {
+      while (atomic_idx_.load() < points_C_.size()) {
+        *point_idx = getMixedIdx(atomic_idx_.fetch_add(1));
```
true, will change

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -155,109 +316,93 @@ class TsdfIntegratorBase {
+    timing::Timer integrate_timer("integrate");
 
-      const Ray unit_ray = (point_G - origin).normalized();
+    point_getter_ =
```
Because it needs to know T_G_C, points, etc. on construction and is constructed new every integration request, not once per integrator object?

mfehr replied:
@helenol  @ZacharyTaylor yes, I get the no-default-constructor thing, but there is clearly no shared ownership with anyone -> unique_ptr :)

helenol replied:
Ahh yes 👍

ZacharyTaylor commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+                      const Colors& colors) {
+    IndexSet index_set;
+
+    point_getter_ =
```
was a member so all threads could access it and was shared because it needs parameters on creation. I will redo it as a variable passed to threads.

ZacharyTaylor commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+      // traced (saves time setting up ray tracer for already used points)
+      AnyIndex block_idx =
+          getGridIndexFromPoint(point_G, layer_->block_size_inv());
+      if (index_set.count(block_idx)) {
```
I believe that this combined with the adding neighbors means that all blocks will be allocated. I need to profile to see how significant the speed gain of this is. If it is not significant I will remove it.

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -155,109 +316,93 @@ class TsdfIntegratorBase {
     BlockHashMapType<std::vector<size_t>>::type clear_map;
 
-    bundleRays(T_G_C, points_C, &voxel_map, &clear_map);
+    bundleRays(&voxel_map, &clear_map);
```
While I have no opinion on whether to refactor PointGetter or not, I agree with Marius on the name obfuscating its purpose. Consider changing it to something that implies it also "stores" the data...

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -406,23 +492,24 @@ class MergedTsdfIntegrator : public TsdfIntegratorBase {
       const BlockHashMapType<std::vector<size_t>>::type& voxel_map,
       const BlockHashMapType<std::vector<size_t>>::type& clear_map,
-      std::queue<VoxelInfo>* voxel_update_queue, size_t tid) {
+      size_t tid) {
```
no need for const for fundamental types in my opinion

mfehr replied:
the const for fundamental types might not be strictly necessary because the params are a copy anyways, so no harm is done by changing them. However having them const would raise a compilation error should you accidentally assign something to them, which helps prevent bugs. :) (see [here, second answer](https://stackoverflow.com/questions/117293/use-of-const-for-function-parameters))

alexmillane replied:
I agree, I read somewhere once something along the lines of "pass by copy arguments are just local variables declared in a special place, so declare them const as you would any other local variable".

helenol commented on voxblox/include/voxblox/integrator/tsdf_integrator.h:
```diff
@@ -447,127 +534,113 @@ class MergedTsdfIntegrator : public TsdfIntegratorBase {
+  static constexpr size_t max_consecutive_ray_collisions_ = 8;
+  static constexpr size_t masked_bits_ = 20;  // 8 mb of ram per tester
+  static constexpr size_t full_reset_threshold = 10000;
+  ApproxHashSet<masked_bits_, full_reset_threshold> approx_start_tester_;
```
at least a comment :D

helenol commented on voxblox/include/voxblox/mesh/mesh_integrator.h:
```diff
@@ -183,11 +185,7 @@ class MeshIntegrator {
-      // TODO(helenol): comment above from open_chisel, but no actual checks
-      // on distance are ever made. Definitely we should skip doing checks of
-      // voxels that are too far from the surface...
+      // Do not extract a mesh here if one of the corner is unobserved.
```
Cheeeeeers thanks!

helenol commented on voxblox/include/voxblox/utils/approx_hash_array.h:
```diff
@@ -0,0 +1,153 @@
+// - Highly inefficient use of memory (allocates 2^N elements)
+// - Cannot disern between two different elements with the same hash
+// - If the hash of two elements have the same first N elements of their hash,
+// only one can be store.
```
stored

helenol commented on voxblox/include/voxblox/utils/approx_hash_array.h:
```diff
@@ -0,0 +1,153 @@
+// Basic container, given in an index and get the element that was stored there.
+// There are 2^unmasked_bits_ elements in the container, which element is
+// returned depends on your hashing function.
+// Uses at least 2^unmaksed_bits * sizeof(StoreElement) bytes of ram
```
masked

helenol commented on voxblox/include/voxblox/utils/approx_hash_array.h:
```diff
@@ -0,0 +1,153 @@
+
+// Acts as a fast and thread safe set, with the serious limitation of both
+// false-negatives and false positives being possible.
+// Uses at least (2^unmaksed_bits + full_reset_threshold) * sizeof(StoreElement)
```
masked

ZacharyTaylor commented:
Ok, I believe I have sorted all the review comments. Its now a lot cleaner and I am quite a bit happier with its layout / performance. @mfehr 

For the block casting for memory pre-allocation I initally ran into issues. Casting every ray was quite expensive (around 100 ms on my dataset) and the fix I had of only casting from unique starting locations and then adding the neighbors of blocks, seemed to have some rare corner cases where it would miss a block.

In the end I swapped to an approach loosely inspired by how infinitam does it. Each thread will try get a voxel from the map, if that block has not been allocated it will instead create a voxel in its own local map and use this instead. Once the integration is done a single thread goes through and adds the contents of all the local maps to the real hashmap. This is slow for the first frame, but otherwise quite efficient (maybe 5-10 ms on same tests).

Other then that changes are:
- Fixed bug where verbose would make mutexes in timers deadlock
- Fixed bug that caused rays to end 1 voxel too soon
- Removed point getter and replaced with a much smaller class that just gets the index, plus an internal function to check if the point is valid.
- Split integrator into a cc file
- Added checks for nullpointers and switched existing DCHECK_NOT_NULLs to DCHECK( != nullptr)
- Threaded mesher
- Added unit tests for approx_hash_array
- Added lots more comments around where the approx_hash_array is used
- Added some benchmarks and images to readme

ZacharyTaylor commented:
Also @mfehr do you have any idea why it is now not building on ubuntu in jenkins? The differences from where it built correctly are very minor and the error is coming from the tango part,

/usr/bin/ld: CMakeFiles/mesh_tango_tsdf.dir/test/mesh_tango_tsdf.cc.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line

ZacharyTaylor commented:
test this please

ZacharyTaylor commented:
test this please

helenol commented:
FYI this PR fixes:
 - the deadlock caused when printing timings introduced in #95
 - segfaults if running multiple (multi-threaded) TSDF servers at once in the TSDF integrator

mfehr commented on voxblox/src/integrator/tsdf_integrator.cc:
```diff
@@ -0,0 +1,567 @@
+  tsdf_voxel->distance =
+      (new_sdf > 0.0) ? std::min(config_.default_truncation_distance, new_sdf)
+                      : std::max(-config_.default_truncation_distance, new_sdf);
+  tsdf_voxel->weight = std::min(config_.max_weight, new_weight);
```
how about doing this capping above when you compute the new weight?

ZacharyTaylor replied:
I was mimicking the behavior already present. I am not sure what is best here, will discuss.

mfehr commented on voxblox/src/integrator/tsdf_integrator.cc:
```diff
@@ -0,0 +1,567 @@
+void SimpleTsdfIntegrator::integratePointCloud(const Transformation& T_G_C,
+                                               const Pointcloud& points_C,
+                                               const Colors& colors) {
+  std::vector<VoxelMap> temp_voxel_storage(config_.integrator_threads);
```
Are you sure this temporary voxel map is worth it? Seems to me there will be a lot of allocating and freeing going on in every integration call.

ZacharyTaylor replied:
If we want to multi-thread then unless we serialize access to the map, its size cannot change while the threads are working on it. This leaves two choices pre-allocate the blocks before each update or using temporary storage for missed blocks and merging them in later. (Or using locks inside the map but this slows things down a lot).

The pre-allocation of blocks for the cow and lady dataset usually takes 50 to 100 ms per frame. With the hack I was doing of early termination I got this down to 10 to 20 ms, but it still had issues. The cost of copying these missed blocks is usually 5 to 10 ms (much more for the first frame but much less for most of the others). 

It could be improved further, but it doesn't seem to be bottle-necking things.

mfehr commented on voxblox/src/integrator/tsdf_integrator.cc:
```diff
@@ -0,0 +1,567 @@
+  BlockHashMapType<std::vector<size_t>>::type voxel_map;
+  // This is a hash map (same as above) to all the indices that need to be
+  // cleared.
+  BlockHashMapType<std::vector<size_t>>::type clear_map;
```
I guess it does make sense to rename BlockHasMapType, because here it clearly referes to voxel indices, right?

mfehr commented on voxblox/test/test_approx_hash_array.cc:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  // test insert and clearing
+  for (size_t i = 0; i < 50; ++i) {
```
50u

ZacharyTaylor commented on voxblox/include/voxblox/integrator/integrator_utils.h:
```diff
@@ -12,25 +12,121 @@
+    }
+    DCHECK(ray_index != nullptr);
+    *ray_index = curr_index_;
+    at_end_ = curr_index_ == end_index_;
```
yeah, it was finishing one too early by immediately returning at_end when it should return it on the next call.

ZacharyTaylor commented on voxblox/src/integrator/tsdf_integrator.cc:
```diff
@@ -0,0 +1,567 @@
+  }
+}
+
+// NOT thread safe
```
If we have a map scope lock then we will need to keep track of how many threads are currently reading or writing to voxels so we can tell when they have all left the map and it is safe to begin operating on it. 

We would also need all these readers to check the lock before moving into the function. This style of reader-writer lock can be difficult to implement without noticeable overhead. C++17 has a std lock type for this, but in C++11 we would need to make our own.

mfehr replied:
Sure, if it is too much overhead, let's not do it. But just FYI we have a reader-writer lock implementation in aslam_cv2: [reader-writer-lock.h](https://github.com/ethz-asl/aslam_cv2/blob/master/aslam_cv_common/include/aslam/common/reader-writer-lock.h) and [reader-first-reader-writer-lock.h](https://github.com/ethz-asl/aslam_cv2/blob/master/aslam_cv_common/include/aslam/common/reader-first-reader-writer-lock.h)
